### PR TITLE
Auto index

### PR DESCRIPTION
**Pull request by jkbonfield**
1. Fixed the travis build so it behaves as claimed in the comment - an *optimised* build doing address sanitizer. 

2. Added a global --write-index option to enable the auto-indexing code.  This requires the auto_index PR of htslib merging first.

### CODE REVIEW

**Comment by jkbonfield** [comment]
NB: fixes #774

**Comment by jkbonfield** [comment]
Note I restarted tests and Travis now passes thanks to the htslib merge, but AppVeyor dies:

> The command failed [256]: C:/projects/samtools/samtools view -h --write-index -O sam,level=5 -o C:/projects/samtools/test/dat/auto_indexed.sam C:/projects/samtools/test/dat/mpileup.1.sam
> out:
> err:samtools auto_index: failed to open index "C:/projects/samtools/test/dat/auto_indexed.sam.csi" for writing

This error comes from sam_utils.c due to `sam_idx_init` returning < 0.    I'm wondering if it's something odd with C: in the filename, but I'm curious how it passes htslib, given that is the bit that does the writing of the index, but not samtools.

Furthermore it passes AppVeyor when run from my own fork: https://ci.appveyor.com/project/jkbonfield/samtools/builds/25278919

I'll restart it incase something broken in the merge, but this has me stumped atm.

**Comment by daviesrob** [comment]
Added support for `##idx##` delimiter that came in with samtools/htslib#861, but see also samtools/htslib#884 which is needed to prevent samtools from incorrectly outputting sam all the time when this is used.

File: bam_addrprg.c [hunk 1]
```diff
@@ -438,12 +438,17 @@ static bool init(const parsed_opts_t* opts, state_t** state_out) {
...
 
-static bool readgroupise(state_t* state)
+static bool readgroupise(parsed_opts_t *opts, state_t* state)
 {
     if (sam_hdr_write(state->output_file, state->output_header) != 0) {
         print_error_errno("addreplacerg", "[%s] Could not write header to output file", __func__);
         return false;
     }
+    char *idx_fn = NULL;
+    if (opts->ga.write_index && opts->output_name && strcmp(opts->output_name, "-") != 0) {
```
Comment (valeriuo) [inline bam_addrprg.c]: `if (opts->ga.write_index)` should be enough, as `auto_index` checks internally for a null output file name and for `stdout`.

File: padding.c [hunk 2]
```diff
@@ -597,10 +597,23 @@ int main_pad2unpad(int argc, char *argv[])
         ret = 1;
         goto depad_end;
     }
+    if (ga.write_index) {
```
Comment (valeriuo) [inline padding.c]: Here, only the flag is checked, as it probably should be.

**Comment by valeriuo** [comment]
Rebased and pushed to develop

**Pull request closed without merging**